### ISSUE TITLE
Exceedingly different outputs on cuda/cpu

### ISSUE THREAD

**Issue by michele-arrival** [pytorch/pytorch#53178]
## 🐛 Bug

I'm running a simple conv2d with padding and dilation, and i notice big difference in results between the cpu and cuda runs (25% of samples are different - i.e. not close enough)

## To Reproduce

```
import torch
from torch import nn
from torch.nn import functional as F
import numpy as np

inps = 2
outs = 512
w, h = 16, 8
ks = 5
b = 4

data_np = np.random.random([b, inps, h, w])
par = nn.Parameter(torch.empty(size=[outs, inps, ks, ks]))
nn.init.kaiming_normal_(par, a=2)

res = {}

for device in ["cpu", "cuda"]:
    tensor = torch.from_numpy(data_np).to(device).to(torch.float32)
    kernel = par.to(device)

    output = F.conv2d(tensor, weight=kernel, padding=(4, 4), dilation=[2, 2])

    res[device] = output

np.testing.assert_allclose(*[r.detach().cpu().numpy() for r in res.values()],
                           rtol=2e-3,
                           atol=1e-3)
```

## Expected behavior

I would expect the outputs to broadly match within the given tolerance (`rtol=1e-1, atol=1e-1`) yet 25% of the output is `consistently` different.

```
AssertionError:
Not equal to tolerance rtol=0.1, atol=0.1

Mismatched elements: 26351 / 262144 (10.1%)
Max absolute difference: 0.67974156
Max relative difference: 11097.508
```

## Environment

PyTorch version: 1.9.0.dev20210302
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: 10.0.0-4ubuntu1
CMake version: version 3.15.4

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: TITAN RTX
GPU 1: TITAN RTX
GPU 2: TITAN RTX

Nvidia driver version: 450.102.04
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.9.0.dev20210302
[pip3] torchaudio==0.8.0a0+5566138
[pip3] torchvision==0.9.0.dev20210302
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] mkl                       2020.2                      256
[conda] mkl-service               2.3.0            py37he8ac12f_0
[conda] mkl_fft                   1.3.0            py37h54f3939_0
[conda] mkl_random                1.1.1            py37h0573a6f_0
[conda] numpy                     1.19.2           py37h54aff64_0
[conda] numpy-base                1.19.2           py37hfa32c7d_0
[conda] pytorch                   1.9.0.dev20210302 py3.7_cuda10.2_cudnn7.6.5_0    pytorch-nightly
[conda] torchaudio                0.8.0.dev20210302            py37    pytorch-nightly
[conda] torchvision               0.9.0.dev20210302      py37_cu102    pytorch-nightly

## 

What am I doing wrong?

**Comment by michele-arrival** [comment]
It seems to be dependent on the number of outputs too, as with `outs = 256` there doesn't seem to be any discrepancy

**Comment by xwang233** [comment]
I've verified this on cuda 10.2 + cudnn 7.6.5. It seems that convolution with padding and dilation was not well supported on cudnn 7.6.5 with this shape (however, cuda 11 + cudnn 8 seems fine). We can disable it from using cudnn < 8, but I'm not sure if the problem is dilation or padding. 

cc @ngimel @ptrblck

**Comment by ngimel** [comment]
Can you try running some tests to establish which parameters are problematic?

**Comment by michele-arrival** [comment]
> Can you try running some tests to establish which parameters are problematic?

Hi @ngimel . What do you mean? There is really only 1 parameter:
```
par = nn.Parameter(torch.empty(size=[outs, inps, ks, ks]))
nn.init.kaiming_normal_(par, a=2)
```

**Comment by xwang233** [comment]
> Can you try running some tests to establish which parameters are problematic?

That's for me. 😅 

@michele-arrival We would suggest upgrading to cuda 11 + cudnn 8 libraries. If that's not possible, you may use a workaround to disable cudnn convolution in your script. For example

```python
with torch.backends.cudnn.flags(enabled=False):
    # your torch.nn.Conv2d code here
```

**Comment by michele-arrival** [comment]
> > Can you try running some tests to establish which parameters are problematic?
> 
> That's for me. 😅
> 
> @michele-arrival We would suggest upgrading to cuda 11 + cudnn 8 libraries. If that's not possible, you may use a workaround to disable cudnn convolution in your script. For example
> 
> ```python
> with torch.backends.cudnn.flags(enabled=False):
>     # your torch.nn.Conv2d code here
> ```

Ehm I think some other rubbish comes up when I move to cuda 11 + cudnn 8. I will post that too!

Disabling cudnn is not really an option, because of the performance penalty. I rather taylor my code to not have such small convolutions.

Never the less i will report on the cu11+cudnn8 issue shortly.

**Comment by michele-arrival** [comment]
Better late than never, these 2 configurations fail with both pt 1.7.1 and 1.8.1 on cuda 11 (11.0 and 11.1 respectively). Neither fails with cuda 10.2 (on pytorch version 1.7.1 or 1.8.1).

```
w, h = 16, 8
inps = 128
outs = 128
padding=(1, 1)  #  anything > 0
dilation=None  #  or (1, 1)
```

```
w, h = 32, 16
inps = 32
outs = 32
padding=(1, 1)  #  anything > 0
dilation=None  #  or (1, 1)
```
with tolerances like before, the values differ less than 0.1% of the times. 
Admittedly not too much, but still enought to raise some flags in my opinion.

with rtol=2e-4, atol=1e-4, the fail rate goes up to 6-10%

You can test with this little snippet:

```
import torch
from torch import nn
from torch.nn import functional as F
import numpy as np

print("TORCH VERSION: {}".format(torch.version.__version__))
print("CUDA VERSION: {}".format(torch.version.cuda))
print("CUDNN VERSION: {}".format(torch.backends.cudnn.version()))

for params in [
    [[4,  2,  8, 16], 512, 5, 4, 2],
    
    [[2, 128, 8, 16], 128, 5, 1, 1],
    [[2, 128, 8, 16], 128, 5, 2, 1],

    [[2, 32, 16, 32],  32, 5, 1, 1],
    [[2, 32, 16, 32],  32, 5, 2, 1],
    ]:
    
    input_size, outs, ks, padding, dilation = params

    data_np = np.random.random(input_size)
    par = nn.Parameter(torch.empty(size=[outs, input_size[1], ks, ks]))
    nn.init.kaiming_normal_(par, a=2)


    res = {}

    for device in ["cpu", "cuda"]:
        tensor = torch.from_numpy(data_np).to(device).to(torch.float32)
        
        kernel = par.to(device).to(torch.float32)
        output = F.conv2d(tensor, weight=kernel, padding=(padding, ) * 2, dilation=(dilation, ) * 2)

        res[device] = output
        # print(device, output.shape)
    
    print(params, end=": ")
    try:
        np.testing.assert_allclose(*[r.detach().cpu().numpy() for r in res.values()],
                                rtol=0.0002,
                                atol=0.0001)

        print("PASS")
    except:
        print("FAIL")
```

While i tested initially with pytorch 1.7.1 and 1.8.1, I report below with nightly:

```
TORCH VERSION: 1.9.0.dev20210326
CUDA VERSION: 10.2
CUDNN VERSION: 7605
[[4, 2, 8, 16], 512, 5, 4, 2]: FAIL
[[2, 128, 8, 16], 128, 5, 1, 1]: PASS
[[2, 128, 8, 16], 128, 5, 2, 1]: PASS
[[2, 32, 16, 32], 32, 5, 1, 1]: PASS
[[2, 32, 16, 32], 32, 5, 2, 1]: PASS
```

```
TORCH VERSION: 1.9.0.dev20210326
CUDA VERSION: 11.1
CUDNN VERSION: 8005
[[4, 2, 8, 16], 512, 5, 4, 2]: PASS
[[2, 128, 8, 16], 128, 5, 1, 1]: FAIL
[[2, 128, 8, 16], 128, 5, 2, 1]: FAIL
[[2, 32, 16, 32], 32, 5, 1, 1]: FAIL
[[2, 32, 16, 32], 32, 5, 2, 1]: FAIL
```

**Comment by michele-arrival** [comment]
Interestingly enough, on Windows, cuda 11 fails on the 2nd and 3rd, but cuda 10.2 doesn't fail at all

```
TORCH VERSION: 1.9.0.dev20210326
CUDA VERSION: 11.1
CUDNN VERSION: 8005
[[4, 2, 8, 16], 512, 5, 4, 2]: PASS
[[2, 128, 8, 16], 128, 5, 1, 1]: FAIL:  Mismatched elements: 1492 / 21504 (6.94%)
[[2, 128, 8, 16], 128, 5, 2, 1]: FAIL:  Mismatched elements: 3172 / 32768 (9.68%)
[[2, 32, 16, 32], 32, 5, 1, 1]: PASS
[[2, 32, 16, 32], 32, 5, 2, 1]: PASS
```

```
TORCH VERSION: 1.9.0.dev20210326
CUDA VERSION: 10.2
CUDNN VERSION: 7605
[[4, 2, 8, 16], 512, 5, 4, 2]: PASS
[[2, 128, 8, 16], 128, 5, 1, 1]: PASS
[[2, 128, 8, 16], 128, 5, 2, 1]: PASS
[[2, 32, 16, 32], 32, 5, 1, 1]: PASS
[[2, 32, 16, 32], 32, 5, 2, 1]: PASS
```

:confused:

**Comment by dennis-park-TRI** [comment]
I'm seeing similar issue: `Conv2d` gives wrong results (compared with cpu version and cpu/gpu version of cuda 10.2). Any update on this?
